File: medium/group-anagrams.go
func groupAnagrams(strs []string) [][]string {
	//     Input: strs = ["eat","tea","tan","ate","nat","bat"]
	//     Output: [["bat"],["nat","tan"],["ate","eat","tea"]]

	//     loop str in strs
	//     loop char in str
	//     get int char and put in array k
	//     put array k in map[[26]int][]string

	//     mapStr := make(map[[26]int][]string)
	//     for _, str := range strs {
	//         k := [26]int{}
	//         for _, character := range str {
	//             k[int(character)-int('a')] += 1
	//         }
	//         mapStr[k] = append(mapStr[k], str)
	//     }

	//     ans := [][]string{}
	//     for _, str := range mapStr {
	//         ans = append(ans, str)
	//     }
	//     return ans

	// using sort
	mapStr := make(map[string][]string)
	for _, str := range strs {
		cloneStr := str
		cloneStr = SortCharInString(cloneStr)
		mapStr[cloneStr] = append(mapStr[cloneStr], str)
	}

	ans := [][]string{}
	for _, str := range mapStr {
		ans = append(ans, str)
	}
	return ans
}

func SortCharInString(str string) string {
	s := strings.Split(str, "")
	sort.Strings(s)
	return strings.Join(s, "")
}

File: medium/group-anagrams.py
#"eat","tea","tan","ate","nat","bat"
#
# create a hash map with a key is lists of character count, values are string
# [1e 1a 1t] eat aet tea
# use an int array with the size of 26 as key

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        res = defaultdict(list) 
        # # mapping charCount to list of Anagrams
        # # defaultdict is a dict, hash map with a default value of an empty list

        for str in strs:
            count = [0]*26
            for ch in str:
                count[ord(ch)-ord("a")] += 1
            res[tuple(count)].append(str)
            # use tuple because tuple is hashable, immutable, it can be used as a key to a hash map
            # list can't be a key in hash map
        return res.values()

        # result = {}
        # for str in strs:
        #     count = [0]*26
        #     for ch in str:
        #         count[ord(ch)-ord("a")] += 1    
        #     if tuple(count) in result:
        #         result[tuple(count)].append(str)
        #     else:
        #         result[tuple(count)] = [str]
        # return result.values()
